SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyAllocStack.swift

```diff
@@ -302,4 +303,37 @@ private extension AllocStackInst {
     }
     return concreteType
   }
+
+  /// Replaces an alloc_stack of an Pack Element by an alloc_stack of the concrete type.
+  ///
+  /// For example:
+  /// ```
+  ///   %0 = alloc_stack $@pack_element(...) ...
+  ///   use %1
+  /// ```
+  /// is transformed to
+  /// ```
+  ///   %0 = alloc_stack $T
+  ///   use %0
+  /// ```
+  ///
+  private func optimizePackElement(_ context: SimplifyContext) -> Bool {
+    guard let concreteType = concreteTypeOfDependentPackElementArchetype else {
+      return false
+    }
+
+    let builder = Builder(before: self, context)
+    let newAlloc = builder.createAllocStack(concreteType.loweredType(in: parentFunction),
+                                            hasDynamicLifetime: hasDynamicLifetime,
+                                            isLexical: isLexical,
+                                            isFromVarDecl: isFromVarDecl,
+                                            usesMoveableValueDebugInfo: usesMoveableValueDebugInfo)    
+    for use in uses {
+        // FIXME: Are there any specific cases to handle?
```

Comment:
yes! You cannot replace a value with a new value which has a different type without looking at the specific using instructions.

The safest thing to do is to cast the new alloc_stack back to the original pack existential type and let other peephole optimizations clean that up:

```
  %1 = alloc_stack $@pack_element(...)
```
->
```
  %0 = alloc_stack $T
  %1 = unchecked_addr_cast %0 to $*@pack_element(...)
```

For example, if an original use is an `unchecked_addr_cast` itself, the resulting double address casts will be optimized away.

However, you might need to add some peephole optimizations for the address cast, e.g. optimize `unchecked_addr_cast` - `witness_method` pairs